Updating HAL/VM ABI to support indirect bindings (COMPATIBILITY BREAKING).

benvanik opened the pull request:
This bumps the HAL module version to 0.3 and makes a few changes required to support reusable/indirect command buffers.

This is the first breaking change in the HAL ABI in 2024 and hopefully the only one required this year!

Summary of changes:
* `hal.command_buffer.create` now takes a set of allowed queue affinities to match `iree_hal_command_buffer_create`
* `hal.command_buffer.update_buffer` was added and maps to `iree_hal_command_buffer_update_buffer`
* `hal.command_buffer.fill_buffer`, `hal.command_buffer.copy_buffer`, `hal.command_buffer.collective`, and `hal.command_buffer.dispatch.indirect` were updated to take binding table slots in addition to buffer pointers

Progress on #17875.

benvanik committed:
Adding `hal.command_buffer.update_buffer`.

benvanik committed:
Adding queue affinity arg to `hal.command_buffer.create`.
The C API has taken this for awhile as a hint to implementations as to
which queues a command buffer may be executed on. It's legal for this to
always be "any" but we may want to scope things more tightly in the
future to e.g. replicate command buffers on multiple logical devices
represented as queues or manage NUMA hinting.

benvanik committed:
Updating HAL VM ABI to pass binding table slots.

benvanik committed:
Bumping HAL module version to 0.3.
This is a breaking change between compiler and runtime and versions
will not be compatible. Compiled artifacts using the HAL must be
recompiled.

ScottTodd reviewed (approved):
LGTM. Timing wise... I think we could land this whenever. We're due for a stable release soon, so we could cut a release _before_ this change then let this soak until the next release?

A few people are working on branches (like https://github.com/iree-org/iree/tree/shared/sdxl_quantized), so they'll want to be careful/intentional about pulling this in.

benvanik committed:
Adding `iree_hal_dispatch_flags_t` to dispatch operations.
This is currently unused but may be useful for specifying scheduling
behavior or something else.

ScottTodd commented on compiler/src/iree/compiler/Dialect/HAL/IR/HALAttrs.td:
```diff
@@ -186,6 +186,14 @@ def HAL_DescriptorSetLayoutFlagsAttr :
+    I64BitEnumAttr<"DispatchFlags", "valid Dispatch flags", [
+      HAL_DispatchFlags_None,
+    ]> {
+  let cppNamespace = "::mlir::iree_compiler::IREE::HAL";
```
nit: why the capitalization in `valid Dispatch flags`?

benvanik replied:
copy/paste :P fixed

ScottTodd reviewed (approved):
LGTM when CI passes (some crashes compiling test programs?)
